### PR TITLE
[Exclusivity] Enforce exclusive access for class offsets in KeyPaths

### PR DESCRIPTION
**Pull request by devincoughlin**
Add dynamic enforcement of exclusive access when a KeyPath directly accesses a final
stored property on an instance of a class. For read-only projections, this begins and ends
the access immediately. For mutable projections, this uses the ClassHolder to perform
a long-term access that lasts as long as the lifetime of the ClassHolder.

rdar://problem/31972680

### CODE REVIEW

File: stdlib/public/core/KeyPath.swift
Comment (atrick) [inline stdlib/public/core/KeyPath.swift]: Why not?

File: stdlib/public/core/KeyPath.swift
Comment (atrick) [inline stdlib/public/core/KeyPath.swift]: I guess this is more a question for @jckarter. Why did we drop down to withUnsafeMutablePointer, instead of just holding on to an Unmanaged reference?

File: test/Interpreter/enforce_exclusive_access_backtrace.swift
Comment (atrick) [inline test/Interpreter/enforce_exclusive_access_backtrace.swift]: typo: kKeyey

**Review by atrick — APPROVED** [review]
Awesome!

File: stdlib/public/core/KeyPath.swift
Comment (devincoughlin) [reply stdlib/public/core/KeyPath.swift]: There is no Swift-level way to initialize a value of UnsafeValueBuffer type, so the initializer doesn't pass DI. I'll update the comment with more information.

File: stdlib/public/core/KeyPath.swift
Comment (jckarter) [reply stdlib/public/core/KeyPath.swift]: I don't see any reason we couldn't use `Unmanaged` and manual reference counting here. Are there undesired interactions between the exclusivity calls and the implicit refcounting at the unmanaged-managed boundaries?

Comment (atrick) [reply stdlib/public/core/KeyPath.swift]: Nevermind, it's obvious now that I look at this again, the ClassHolder is now allocated with a builtin. There's just no way to get the memory in an initialized state without using an unsafe pointer.

**Comment by devincoughlin** [comment]
Updated to add additional comments about allocation and initialization of ClassHolder instances. Also rebased on top of master.

**Comment by devincoughlin** [comment]
@swift-ci Please smoke test and merge

**Comment by devincoughlin** [comment]
Rebase on top of @atrick 's recent 'noNestedConflict' changes. In SILGenBuiltin.cpp I've made emitBuiltinBeginUnpairedModifyAccess() have 'noNestedConflict' set to false and emitBuiltinPerformInstantaneousReadAccess() have it set to true. @atrick does this seem right to you?

**Comment by devincoughlin** [comment]
@swift-ci please smoke test

**Comment by atrick** [comment]
@devincoughlin that looks good to me.

**Comment by atrick** [comment]
@swift-ci smoke test OS X Platform.

**Comment by atrick** [comment]
@devincoughlin you can either set noNestedAccess to false and let the future optimization handle it, or wait until this PR is merged and retest: https://github.com/apple/swift/pull/15659

**Comment by atrick** [comment]
@swift-ci smoke test.

**Comment by devincoughlin** [comment]
@swift-ci Please smoke test

**Comment by atrick** [comment]
```
Interpreter/enforce_exclusive_access.swift

19:58:09 [ RUN      ] ExclusiveAccess.KeyPathInoutDirectWriteClassStoredProp
19:58:09 expecting a crash, but the test did not crash
19:58:09 [     FAIL ] ExclusiveAccess.KeyPathInoutDirectWriteClassStoredProp
19:58:09 [ RUN      ] ExclusiveAccess.KeyPathInoutDirectReadClassStoredProp
19:58:09 expecting a crash, but the test did not crash
19:58:09 [     FAIL ] ExclusiveAccess.KeyPathInoutDirectReadClassStoredProp

```
So, the IRGen "fix" I mentioned above,
https://github.com/apple/swift/pull/15659, should not have changed any functionality. Skipping the endAccess runtime call is purely an optimization (which is why I originally lumped it in with my
optimization pass). If the beginAccess isn't tracked, then the scratch buffer will be invalid and endAccess will do nothing.

What's more, I don't think the no_nested_conflict changes should have any effect on the KeyPathInoutDirectWriteClassStoredProp test that is failing (there's no instantaneous access in that test).

So, was the test ever passing? I looked into it and noticed that it requires the standard library to emit the unpaired access markers. We had never intended to enable access markers in the standard library because, once it's been tested, there's no point shipping it with extra checks. This is generally the right strategy, but doesn't work for the Builtin checks, which actually form an outer access scope surrounding user code.

I'm tempted to suggest handling the Builtins specially, and unconditionally emitting markers. But first we should consider whether there are other instances of this issue. Maybe the solution needs to
be generalized. As long as the stdlib doesn't instantiate internal classes with materializeForSet properties, I don't *think* this problem will show up elsewhere. @rjmccall can you think of any other places that we need to emit stdlib checks to pick up on user violations?

**Comment by devincoughlin** [comment]
Oof. That explains why the test passes locally. I'm using a debug version of the standard library.

**Comment by rjmccall** [comment]
I would hope that the standard library generally doesn't need dynamic checks because it can just use static enforcement.  So the approach I would suggest would be to look at all the dynamic checks that we *do* emit and either (1) decide they're statically okay (and disable them with an attribute?) or (2) leave them in as dynamic checks.  A base policy of disabling all the checks feels like going the wrong way about it.

**Comment by atrick** [comment]
@gottesmm plans to add a function-level `@semantics` attribute to disable access marker elimination (i.e. force them to be emitted). That should unblock this fix.
<rdar://problem/39335800> Add a function attribute to KeyPath to force exclusivity enforcement.

**Comment by gottesmm** [comment]
@devincoughlin I am going to chop off a part of this commit since I need it to test my semantics thing. Specifically I am just going to cut off the builtin emission part.

**Comment by atrick** [comment]
@devincoughlin, Since @gottesmm is testing this in conjunction with his changes, he'll probably end of staging it back in as two parts it in new PRs.

**Pull request closed without merging**
